README.md
# One-Date-One-Question
this is a space for self-improvement of mine

### 工程启动步骤

#### 安装所需依赖
npm / cnpm install 

#### 启动热更新服务
npm run server  

#### 访问地址
http://127.0.0.1:8080/index.html


#### 打包
npm run build

### 历史题目入口

> [历史入口](https://github.com/dorseysen/One-Date-One-Question/blob/master/src/history.md)  

现对新加的每道题目加入难度等级。其中：★为完整一颗星，代表难度等级2，☆为半颗星，代表难度1



src/history.md

### 2019年4月
[本月入口](https://github.com/dorseysen/One-Date-One-Question/blob/master/src/2019/2019-04.js)

> 2019-04-28：[1, 2, 3, 50, 88, 65, 4, 89, 1111]，输出[88, 89, 1111]。 （大于80的所有数），至少两种解法。  

> 2019-04-29： 不通过sort方法，对任意数组（成员都是整数）进行排序，如[9, -2, 200, 35]，输出：[-2, 9, 35, 200];  

> 2019-04-30： 有数组：[1, 2, 100, 80, 33, 90] 和 [-8, 77, 90, 34, 56, 1179]，求中位数。  

### 2019年5月
[本月入口](https://github.com/dorseysen/One-Date-One-Question/blob/master/src/2019/2019-05.js)

#### 五一至五四劳动节休息

> 2019-05-05：请写一个函数，用于判断类型。比如[1]返回array，1返回number，'1'返回string。  

> 2019-05-06：JSON数据排序。  

> [2019-05-07：当a的值在什么的情况下，可以使下列式子打印出1：if (a == 1 && a == 2 && a == 3) {console.log(1);}](https://github.com/dorseysen/One-Date-One-Question/issues/1)  

> [2019-05-08：有字符串'dhaoshdaoidhadia'，打印出最长不重复的字符串片段。](https://github.com/dorseysen/One-Date-One-Question/issues/2)  

> [2019-05-09：请写一个获取url参数的函数，将所有参数转化成key-value键值对](https://github.com/dorseysen/One-Date-One-Question/issues/3)  

> [2019-05-10：写一个数组去重函数](https://github.com/dorseysen/One-Date-One-Question/issues/4)  

> [2019-05-11：统计一个字符串中出现最多的字母（注意是字母），并输出这个次数](https://github.com/dorseysen/One-Date-One-Question/issues/5)  

> [2019-05-12：不借助临时变量，进行两个数的交换](https://github.com/dorseysen/One-Date-One-Question/issues/6)  

> [2019-05-13：找出某个数组最大的差值，比如[1, 2, 8, 23, 4]，最大的差值为23 - 1 = 22](https://github.com/dorseysen/One-Date-One-Question/issues/7)  

> [2019-05-14：输出某个范围内的素数](https://github.com/dorseysen/One-Date-One-Question/issues/8)  

> [2019-05-15：把一个数组arr按照指定的数组大小size分割成若干个数组块,比如输入是[1,2,3,5,4]和2，输出[[1,2],[3,5],[4]]](https://github.com/dorseysen/One-Date-One-Question/issues/9)  

> [2019-05-16：判断一个字符串是否是回文字符串](https://github.com/dorseysen/One-Date-One-Question/issues/10)  

> [2019-05-17：写一个简单的数据深拷贝，纯数据](https://github.com/dorseysen/One-Date-One-Question/issues/11)  

> [2019-05-18：解释一下什么是短路、断路操作符](https://github.com/dorseysen/One-Date-One-Question/issues/12)  

> [2019-05-19：JavaScript写一个获取某范围内随机整数的函数，可随意改变这个范围](https://github.com/dorseysen/One-Date-One-Question/issues/13)  

> [2019-05-20：今天是520，随意说说用JavaScript表达爱意的一种方式，最好是贴出你的代码。](https://github.com/dorseysen/One-Date-One-Question/issues/14)  

> [2019-05-21：手写一个生成UUID字符串的函数](https://github.com/dorseysen/One-Date-One-Question/issues/15)  

> [2019-05-22：求一个数组中，最大与最小数之差](https://github.com/dorseysen/One-Date-One-Question/issues/16)  

> [2019-05-23：输入：abcd123efg 输出 efg123abcd](https://github.com/dorseysen/One-Date-One-Question/issues/17)  

> [2019-05-24：输入yyyy-MM-dd，输出当前时间的年月日，格式根据输入来，假如输入换成yyyy年MM月dd日，输出得自然更改格式](https://github.com/dorseysen/One-Date-One-Question/issues/18)  

> [2019-05-25：hello-my-name-is-dorsey-sen变为驼峰式的：helloMyNameIsDorseySen](https://github.com/dorseysen/One-Date-One-Question/issues/19)  

> [2019-05-26：多个数组合并，并扁平化，同时去重](https://github.com/dorseysen/One-Date-One-Question/issues/20)  

> [2019-05-27：计算一个正整数的阶乘，比如输入：4，则输出：1*2*3*4 = 24。](https://github.com/dorseysen/One-Date-One-Question/issues/21)  

> [2019-05-28：斐波那契数列。即：1、1、2、3、5、8、13、21...求第N个数，输入正整数，如输入是8，输出是21，并打印该完整数列](https://github.com/dorseysen/One-Date-One-Question/issues/22)  

> [2019-05-29：随机生成指定长度字符串。如输入8，则随意输出abc12dfe，字符范围限制在字母与数字。](https://github.com/dorseysen/One-Date-One-Question/issues/23)  

> [2019-05-30：统计一个字符串中出现最多的字符，并返回该字符的次数，同时返回剔除掉该字符后的字符串。如：输入'dhaosdhaincapdnaaaaa'，输出：[8, 'dhosdhincpdn']](https://github.com/dorseysen/One-Date-One-Question/issues/24)  

> [2019-05-31：实现在10 ~ 100这个范围内随机取10个数，存入数组，并做排序，需要注意可能会出现重复的情况](https://github.com/dorseysen/One-Date-One-Question/issues/24)  

### 2019年6月
[本月入口](https://github.com/dorseysen/One-Date-One-Question/blob/master/src/2019/2019-06.js)

> [ 2019-06-01：今天是六一儿童节，老师带领班上的小朋友表演节目，但小朋友们需要鼓励，所以老师准备了很多块巧克力来作为奖品，巧克力有大有小，小朋友愿意上台表演的最小鼓励也是有高有低，而且巧克力既不可以多块给同一人，也不能掰开分给多个小朋友。已知巧克力数和小朋友数目一致，请问如何让尽可能多的小朋友上台表演的前提下，分出去尽可能多（总重量最大）的巧克力？比如说，有重量为[1, 1.2, 1.2, 1.3, 1, 1.5, 2.5, 1.4, 1.6, 2]的巧克力和要求为[1.3, 1.1, 1.2, 1, 2, 1.4, 1.6, 1.7, 1.3, 1.5]的小朋友](https://github.com/dorseysen/One-Date-One-Question/issues/25)  

> [2019-06-02：约瑟夫环问题](https://github.com/dorseysen/One-Date-One-Question/issues/26)  

> [2019-06-03：打印接下来的N个闰年，若今年即为闰年则从今年开始打印](https://github.com/dorseysen/One-Date-One-Question/issues/27)  

> [2019-06-04： 找到一个英文句子中最长的单词，并输出这个单词和单词长度，当有多个最长单词时允许只输出其中一个。如输入：I am a boy with a dream, who want to make some achievements in software industry, little but clinging. 输出：['achievements', 12]](https://github.com/dorseysen/One-Date-One-Question/issues/28)  

> [2019-06-05：写一个函数，用于计算在当下时间前几个月是什么时候。](https://github.com/dorseysen/One-Date-One-Question/issues/29)  

> [2019-06-06：求某一范围内的水仙花数，水仙花数的定义是：某数数位的立方和等于它本身，如：1³ + 5³ + 3³ = 153，153就是水仙花数](https://github.com/dorseysen/One-Date-One-Question/issues/30)  

> [2019-06-07：恺撒密码](https://github.com/dorseysen/One-Date-One-Question/issues/31)  

> [2019-06-08：删除数组中所有假值，如：null、undefined、''、false、0、NaN](https://github.com/dorseysen/One-Date-One-Question/issues/32)  

> [2019-06-09：写一个深拷贝，不可以是纯数据的，这时候的拷贝体可能有function](https://github.com/dorseysen/One-Date-One-Question/issues/33)  

> [2019-06-10：踩地雷。一队兵，排成一排，通过某个雷区，雷区间隔一定距离埋入地雷，假如间隔刚好5个人，问此时能安全通过雷区的还剩下谁。比如：1-30号的士兵，此时从1号开始间隔5个人站位有地雷，此时：1-5号都是安全的，但6号会被炸，7-11号也是安全的，12号被炸](https://github.com/dorseysen/One-Date-One-Question/issues/34)  

> [2019-06-11：输入两个数组，返回两个数组中独有的值。如：[1, 2, 3] 和 [2, 3, 4, 5]。返回 [1] 及 [4, 5]，因为其中2和3不是独有的，去除掉](https://github.com/dorseysen/One-Date-One-Question/issues/35)  

> [2019-06-12：截断一个字符串！如果字符串的长度比指定的参数num长，则把多余的部分用...来表示。值得注意的是，插入到字符串尾部的三个点号也会计入字符串的长度。如果指定的参数num小于或等于3，则添加的三个点号不会计入字符串的长度。](https://github.com/dorseysen/One-Date-One-Question/issues/36)  

> [2019-06-13：判断字符串以字母开头，后面可以是数字，下划线，字母，长度为6-30](https://github.com/dorseysen/One-Date-One-Question/issues/37)  

> [2019-06-14： 找回缺失的数字，假设在一个1-100的数组中，缺失了某个数字，该数组未做排序，请问如何知晓缺失的是什么数。](https://github.com/dorseysen/One-Date-One-Question/issues/38)  

> [2019-06-15：混乱数组最大两个数之和。](https://github.com/dorseysen/One-Date-One-Question/issues/39)  


2019/2019-05.js
//  2019-05-05： 请写一个函数，用于判断类型。比如[1]返回array，1返回number，'1'返回string。

//  the first solution

const getType_20190505_0 = obj => {

    let typeStr = Object.prototype.toString.call(obj);

    return typeStr.split(' ')[1].split(']')[0].toLowerCase();

}
console.log(getType_20190505_0([1]));
console.log(getType_20190505_0(1));
console.log(getType_20190505_0('1'));

//  the second solution
const getType_20190505 = obj => Object.prototype.toString.call(obj).match(/(?<=\s).*?(?=\])/g)[0].toLowerCase();

console.log(getType_20190505([1]));
console.log(getType_20190505(1));
console.log(getType_20190505('1'));

// 2019-05-06： 有JSON数据如下：
// [
//     {
//         TIME: '2019-03-02 23:16:08',
//         records: 'xiaoming'
//     },
//     {
//         TIME: '2019-06-05 20:16:08',
//         records: 'xiaohua'
//     },
//     {
//         TIME: '2019-03-02 23:16:10',
//         records: 'xiaohong'
//     },
// ]
// 请对该数组按时间进行排序。

let arr20190506 = [
    {
        TIME: '2019-03-02 23:16:08',
        records: 'xiaoming'
    },
    {
        TIME: '2019-06-05 20:16:08',
        records: 'xiaohua'
    },
    {
        TIME: '2019-03-02 23:16:10',
        records: 'xiaohong'
    }
];
const arrSort20190506 = arr => {

    let median = {},
        sortArr = [];

    arr.map(item => {
        let time = new Date(item.TIME);
        sortArr.push(time);
        median[time] = item;
    });

    sortArr.sort((a, b) => a - b);

    return sortArr.map(item => {
        return median[item];
    })
}
console.log(arrSort20190506(arr20190506));

//  2019-05-07:  当a的值在什么的情况下，可以使下列式子打印出1：
// if (a == 1 && a == 2 && a == 3) {
//     console.log(1);
// }
let a = {
    val: 1,
    valueOf: function () {
        return a.val ++;
    }
}
if (a == 1 && a == 2 && a == 3) {
    console.log(1);
}

//  2019-05-08： 有字符串'dhaoshdaoidhadia'，打印出最长不重复的字符串片段。

// the first solution
const longestStr_20190508_00 = str => {

    let res = '', medianArr = [];

    for (let i = 0; i < str.length; i++) {

        if (medianArr.indexOf(str[i]) > -1) {
            
            res = res.length > medianArr.length ? res : medianArr.join('');
            i -= (medianArr.length - 1);

            medianArr = [];
            
        } else {
            medianArr.push(str[i]);
        }
    }
    return res;
}
console.log(longestStr_20190508_00('wawawawadorseyhahaha'));

// the second solution
const longestStr_20190508_01 = str => {

    let res = '', median = '';

    for (let i = 0; i < str.length; i++) {
        
        let index = median.indexOf(str[i]);

        index > -1 ? median = median.substr(- index) : median += str[i];
        
        res = res.length < median.length ? median : res;
    }
    return res;
}
console.log(longestStr_20190508_01('wawawawadorseyhahaha'));

//  2019-05-09：请写一个获取url参数的函数，将所有参数转化成key-value键值对
const getUrl_20190509 = () => {

    let url = window.location.href,
        res = {},
        arrUrl = url.match(/(?<=[\?&]).*?(?=[&]|$)/g);
    
    arrUrl && arrUrl.map(item => {
        res[item.split('=')[0]] = item.split('=')[1];
    });
    return res;
}

console.log(getUrl_20190509());

//  2019-05-10：写一个数组去重函数

const removeRepeat_20190510 = arr => {

    return Array.from(new Set(arr));
}

console.log(removeRepeat_20190510([1, 3, 25, 3, 45]));

//  2019-05-11：统计一个字符串中出现最多的字母（注意是字母），并输出这个次数

const theMostLetter_20190511 = str => {

    let map = {},
        res = 0;

    for ( let i = 0; i < str.length; i++ ) {

        /[a-zA-Z]/.test(str[i]) && ( map[str[i]] = 'undefined' !== typeof map[str[i]] ? map[str[i]] + 1 : 1 );
    }

    for ( let i in map ) {
        
        res = res > map[i] ? res : map[i];
    }
    return res;
}

console.log(theMostLetter_20190511('daishddao123456'));

//  2019-05-12：不借助临时变量，进行两个数的交换

const swapData_20190512 = (a, b) => {

    a = a + b;
    b = a - b;
    a = a - b;

    return [a, b];
}
console.log(swapData_20190512(24, 5));

//  2019-05-13：找出某个数组最大的差值，比如[1, 2, 8, 23, 4]，最大的差值为23 - 1 = 22

const maxDiff_20190513 = arr => {

    let max, min;

    arr.forEach(item => {
        max = max > item ? max : item;
        min = min < item ? min : item;
    });

    return max - min;
}
console.log(maxDiff_20190513([1, 2]));

//  2019-05-14：输出某个范围内的素数

const getPrimeNumber_20190514 = (min, max) => {

    let res = [];
    for ( let i = min; i <= max; i ++ ) {
        
        for ( var k = 2; k <= Math.floor(i / 2); k ++ ) {
            
            if ( i % k === 0 ) break;
        }

        if ( k === ( Math.floor(i / 2) + 1 ) ) res.push( i );
    }
    return res;
}
console.log(getPrimeNumber_20190514(1, 20));

//  2019-05-15：把一个数组arr按照指定的数组大小size分割成若干个数组块,比如输入是[1,2,3,5,4]和2，输出[[1,2],[3,5],[4]]
const sliceArr_20190515 = (arr, size) => {

    // console.log(arr.slice(2, 10));
    let res = [],
        i = 0;
    while (i < arr.length) {
        res.push(arr.slice(i, i + size));
        i += size;
    }
    return res;
}
console.log(sliceArr_20190515([2, 3, 4, 5, 6, 7, 8, 9, 0, 10], 3));

//  2019-05-16：判断一个字符串是否是回文字符串

const palindromeString_20190516 = str => str.split('').reverse().join('') === str;

console.log(palindromeString_20190516('aba'));

//  2019-05-17：写一个简单的数据深拷贝，纯数据

const deepCopy_20190517 = obj => JSON.parse(JSON.stringify(obj));

console.log(deepCopy_20190517({ name: 'dorsey', age: '25' }));

//  2019-05-18：解释一下什么是短路、断路操作符

// 短路断路操作符其实就是利用 &&/||/! 等逻辑运算符的熔断机制，达到相应的判断目的。这是一个较好的替代if/else机制的方案。
// 如：

const shortout_20190518 = res => res && res.length && res.map( (item, index) => { index: item } );

//  这里就是运用短路操作符做判空处理，避免报空指针错误。

//  2019-05-19：JavaScript写一个获取某范围内随机整数的函数，可随意改变这个范围

const randomInt_20190519 = (min, max) => Math.floor(Math.random() * (max - min) + min);

//  2019-05-20：今天是520，随意说说用JavaScript表达爱意的一种方式，最好是贴出你的代码。

console.log('你问我爱你有多深 我爱你有几分 我的情也真 我的爱也真 月亮代表我的心');

console.log('You ask how deep I love you I love you how much my love is true my love is true the moon stands for my heart')

const loveYou = () => {

    let theMoon, myLove;

    const youAskMe = 'how deep I love you';
    
    const thenYouAskMe = 'how much I love you';

    myLove = true;
    myLove === true;

    theMoon = 'my heart';

    return 'forever';
}
loveYou();
//  2019-05-21：手写一个生成UUID字符串的函数

const getUUID = () => new Date().getTime().toString(32) + Math.floor((Math.random() * 999999)).toString(32);

console.log(getUUID());

//  2019-05-22：求一个数组中，最大与最小数之差

const differenceVal = arr => Math.max.apply(Math, arr) - Math.min.apply(Math, arr);

console.log(differenceVal([1, 2, 56, 88, 12, 4, -1, -21, -24.1, 56.4]));

//  2019-05-23：输入：abcd123efg 输出 efg123abcd
const changeBlockLocation = str => str.replace(/(^[a-zA-Z]+)(\d+)([a-zA-Z]+$)/g, '$3$2$1');

console.log(changeBlockLocation('abcd123efg'));

//  2019-05-24：输入yyyy-MM-dd，输出当前时间的年月日，格式根据输入来，假如输入换成yyyy年MM月dd日，输出得自然更改格式

const dateFormat = (reg) => {

    let date = new Date();

    return reg.replace('yyyy', date.getFullYear())
        .replace('MM', numFormat( date.getMonth() + 1 ))
        .replace('dd', date.getDate());
}
const numFormat = num => num < 10 ? '0' + num : num;


console.log(dateFormat('yyyy-MM-dd'));
